File: customview/src/main/java/com/bksx/customview/customview/CameraView.java
package com.bksx.customview.customview;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Camera;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Point;
import android.util.AttributeSet;
import android.view.View;

import com.bksx.customview.R;

public class CameraView extends View {
    private final int width;
    private final int height;
    Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    Bitmap bitmap;
    Point point1 = new Point(200, 100);
    Point point2 = new Point(600, 200);
    Matrix matrix = new Matrix();
    Camera camera = new Camera();

    {
        bitmap = BitmapFactory.decodeResource(getResources(), R.mipmap.jump);
        width = bitmap.getWidth() / 2;
        height = bitmap.getHeight() / 2;
    }


    public CameraView(Context context) {
        super(context);
    }

    public CameraView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
//        camera.save();
//        matrix.reset();
//        camera.rotateX(30);
//        camera.getMatrix(matrix);
//        camera.restore();
//        matrix.preTranslate(-(200 + width), -(0 + height));
//        matrix.postTranslate(200 + width, 0 + height);
//
//        canvas.save();
//        canvas.concat(matrix);
//        canvas.drawBitmap(bitmap, 200, 0, paint);
//        canvas.restore();
/////////////////////////////////////////
        canvas.save();
        camera.save();
        camera.rotateX(-10);


        camera.applyToCanvas(canvas);
        camera.restore();

        canvas.drawBitmap(bitmap, point1.x, point1.y, paint);
        canvas.restore();
    }


    //TODO 1.利用camera做平移，上方为y轴正方向,将camera的Matrix传给了canvas使得canvas的坐标系也发生了移动
    /*

        canvas.save();
        camera.save();
        camera.translate(0,200,0);


        camera.applyToCanvas(canvas);
        camera.restore();

        canvas.drawBitmap(bitmap, point1.x, point1.y, paint);
        canvas.restore();

     */

    //TODO 2.matrix，下方为y轴正方向
    /*

        canvas.save();
        matrix.reset();

        matrix.postTranslate(0, 200);
        canvas.concat(matrix);
        canvas.drawBitmap(bitmap, point1.x, point1.y, paint);
        canvas.restore();

     */

    //TODO 3.sample12 第一部分分析
    /*
        camera.save();
        matrix.reset();              TODO 拿到一个初始单列矩阵
        camera.rotateX(30);
        camera.getMatrix(matrix);    TODO 计算当前状态下矩阵对应的状态，并将计算后的矩阵赋值给参数matrix，即-将x轴旋转30度后的结果放到matrix
        camera.restore();
        matrix.preTranslate(-center1X, -center1Y);
        matrix.postTranslate(center1X, center1Y); TODO 两次操作化matrix再变化
        canvas.save();
        canvas.concat(matrix);TODO 将matrix再传给canvas画出相应图形
        canvas.drawBitmap(bitmap, point1.x, point1.y, paint);
        canvas.restore();
     */
}

File: hencoderplus/src/main/java/com/example/hencoderplus/customview/CameraView.kt
package com.example.hencoderplus.customview

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View

class CameraView (context: Context?, attrs: AttributeSet?) : View(context, attrs){

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var bitmap: Bitmap?
    private val camera = Camera()

    init {
        paint.style = Paint.Style.FILL
        paint.strokeWidth = 10F
        paint.color = Color.BLUE
        bitmap = getHeadImg(resources, WIDTH.toInt())
        camera.rotateX(30F)//camera的照射，即旋转中心为canvas坐标原点
        camera.setLocation(0F,0F,-4 * resources.displayMetrics.density)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

//        canvas?.clipRect(0F,0F, OFFSET + WIDTH / 2, OFFSET + WIDTH / 2) 裁切



        //翻转的下半部分
        canvas?.save()
        canvas?.translate(OFFSET + WIDTH / 2,OFFSET + WIDTH / 2)//camera的照射旋转中心也随之移动到了中心
        camera.applyToCanvas(canvas)//先旋转坐标系，再画图得到的就是反转的，注意先旋转
        canvas?.clipRect(-WIDTH / 2, 0F,WIDTH / 2, WIDTH / 2)
        canvas?.drawBitmap(bitmap!!, -WIDTH / 2, -WIDTH / 2,paint)
        canvas?.translate(-(OFFSET + WIDTH / 2),-(OFFSET + WIDTH / 2))
        canvas?.restore()

        //不翻转的上半部分
        canvas?.save()
        canvas?.translate(OFFSET + WIDTH / 2,OFFSET + WIDTH / 2)
        canvas?.clipRect(-WIDTH / 2, -WIDTH / 2,WIDTH / 2, 0F)
        canvas?.drawBitmap(bitmap!!, -WIDTH / 2, -WIDTH / 2,paint)
//        canvas?.drawBitmap(bitmap!!, OFFSET, OFFSET,paint)
        canvas?.translate(-(OFFSET + WIDTH / 2),-(OFFSET + WIDTH / 2))
        canvas?.restore()


    }

    companion object {
        val WIDTH = dp2px(200F)
        val OFFSET = dp2px(100F)

    }
}